File: apps/youtube/extensions/spoof/src/main/java/app/reseam/youtube/spoof/AccountCredentialsInvalidText.java
// SPDX-FileCopyrightText: 2026 AunAli K. <hello@auna.li>
// SPDX-FileCopyrightText: 2026 ReVanced contributors
// SPDX-License-Identifier: GPL-3.0-or-later

package app.reseam.youtube.spoof;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;

import app.reseam.youtube.core.Logger;
import app.reseam.youtube.core.Settings;
import app.reseam.youtube.core.YouTubeContext;

/** Explains the MicroG account error when connectivity proves that the stock text is misleading. */
public final class AccountCredentialsInvalidText {
    private AccountCredentialsInvalidText() {}

    public static String getOfflineNetworkErrorString(String original) {
        if (!Settings.getBoolean("account_credentials_invalid_text", true)) return original;
        try {
            ConnectivityManager manager = (ConnectivityManager) YouTubeContext.get()
                    .getSystemService(Context.CONNECTIVITY_SERVICE);
            NetworkInfo network = manager == null ? null : manager.getActiveNetworkInfo();
            if (network != null && network.isConnected()) {
                Logger.debug(() -> "Network appears to be online, but app is showing offline error");
                int id = YouTubeContext.get().getResources().getIdentifier(
                        "microg_offline_account_login_error", "string",
                        YouTubeContext.get().getPackageName());
                if (id != 0) return "\n" + YouTubeContext.get().getString(id);
            } else {
                Logger.debug(() -> "Network is offline");
            }
        } catch (Exception exception) {
            Logger.error(() -> "getOfflineNetworkErrorString failure: " + exception);
        }
        return original;
    }
}

File: apps/youtube/patch/src/main/kotlin/app/reseam/patches/youtube/misc/AccountCredentialsInvalidText.kt
// SPDX-FileCopyrightText: 2026 AunAli K. <hello@auna.li>
// SPDX-FileCopyrightText: 2026 ReVanced contributors
// SPDX-License-Identifier: GPL-3.0-or-later

package app.reseam.patches.youtube.misc

import app.reseam.patch.Type
import app.reseam.patch.after
import app.reseam.patch.methods
import app.reseam.patch.patch
import app.reseam.patch.points
import app.reseam.patch.settings.section
import app.reseam.patches.youtube.core.YOUTUBE
import app.reseam.patches.youtube.core.YouTubeSettings
import app.reseam.patches.youtube.core.YouTubeSettingsPages
import app.reseam.patches.youtube.core.youTubeSettings

val accountCredentialsInvalidText = patch("Account credentials invalid text") {
    description("Explains the misleading offline account error when the network is available.")
    compatibleWith(YOUTUBE)
    dependsOn(youTubeSettings)
    settings(youTubeSettings, section(YouTubeSettingsPages.Advanced, "Account compatibility", YouTubeSettings.accountCredentialsInvalidText))

    execute {
        resources.addString(
            "microg_offline_account_login_error",
            "If you recently changed your account login details, then uninstall and reinstall MicroG.",
        )
        val offlineBody = resources.id("string", "offline_no_content_body_text_not_offline_eligible")?.toLong()
            ?: error("string/offline_no_content_body_text_not_offline_eligible is missing")

        val texts = methods("offline account error") { literals(offlineBody) }.points {
            invokeVirtual { name("getString"); params(Type.Int); returns(Type.String) }
            argument(1) { literal(offlineBody) }
        }.all
        check(texts.isNotEmpty()) { "No offline account error text" }
        texts.forEach { site ->
            site.next { resultOf(Type.String) }.captureAs("offlineError", Type.String).after {
                capture("offlineError").assign(
                    call(AccountCredentialsInvalidText.getOfflineNetworkErrorString, capture("offlineError")),
                )
            }
        }
    }
}
